File: LeetCode/dp/72.编辑距离.cpp
/*
 * @lc app=leetcode.cn id=72 lang=cpp
 *
 * [72] 编辑距离
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    int dp[505][505];
    int minDistance(string word1, string word2) {
        for(int i = 0;i <= word1.size(); i++){
            dp[i][0] = i;
        }
        for(int j = 0;j <= word2.size(); j++){
            dp[0][j] = j;
        }
        for(int i = 1;i <= word1.size(); i++){
            for(int j = 1;j <= word2.size(); j++){
                if(word1[i-1] == word2[j-1])
                    dp[i][j] = dp[i-1][j-1];
                else{
                    dp[i][j] = min(dp[i-1][j-1],dp[i-1][j]);
                    dp[i][j] = min(dp[i][j], dp[i][j-1]) +1;
                }
            }
        }
        return dp[word1.size()][word2.size()];
    }
};
// @lc code=end
int main()
{
    Solution s;
    cout << s.minDistance("intention", "execution") << endl;
    return 0;
}

File: LeetCode/搜索/72.编辑距离.py
#
# @lc app=leetcode.cn id=72 lang=python3
#
# [72] 编辑距离
#
# https://leetcode.cn/problems/edit-distance/description/
#
# algorithms
# Medium (62.88%)
# Likes:    3454
# Dislikes: 0
# Total Accepted:    523.6K
# Total Submissions: 830.8K
# Testcase Example:  '"horse"\n"ros"'
#
# 给你两个单词 word1 和 word2， 请返回将 word1 转换成 word2 所使用的最少操作数  。
# 
# 你可以对一个单词进行如下三种操作：
# 
# 
# 插入一个字符
# 删除一个字符
# 替换一个字符
# 
# 
# 
# 
# 示例 1：
# 
# 
# 输入：word1 = "horse", word2 = "ros"
# 输出：3
# 解释：
# horse -> rorse (将 'h' 替换为 'r')
# rorse -> rose (删除 'r')
# rose -> ros (删除 'e')
# 
# 
# 示例 2：
# 
# 
# 输入：word1 = "intention", word2 = "execution"
# 输出：5
# 解释：
# intention -> inention (删除 't')
# inention -> enention (将 'i' 替换为 'e')
# enention -> exention (将 'n' 替换为 'x')
# exention -> exection (将 'n' 替换为 'c')
# exection -> execution (插入 'u')
# 
# 
# 
# 
# 提示：
# 
# 
# 0 <= word1.length, word2.length <= 500
# word1 和 word2 由小写英文字母组成
# 
# 
#

# @lc code=start
from functools import cache


class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n = len(word1)
        m = len(word2)
        
        @cache
        def dfs(i, j):
            if i < 0 :
                return j + 1
            if j < 0:
                return i + 1
            if word1[i] == word2[j]:
                return dfs(i-1, j-1)
            return min(dfs(i-1, j), dfs(i, j-1), dfs(i-1, j-1)) + 1
        
        return dfs(n-1, m-1)    
# @lc code=end

